File: lib/tauri/ai.ts
import { invoke } from '@tauri-apps/api/core';
import type {
  WritingTask,
  StructuredOutput,
  GenerationRun,
  StyleProfile,
  StyleExample,
} from './types';

export type { WritingTask, StructuredOutput, GenerationRun, StyleProfile, StyleExample };

// ── AI Generation ──

export async function assembleAiPrompt(
  noteId: number,
  task: WritingTask,
): Promise<string> {
  return await invoke<string>('assemble_ai_prompt', { note_id: noteId, task });
}

export async function parseAiOutput(rawOutput: string): Promise<StructuredOutput> {
  return await invoke<StructuredOutput>('parse_ai_output', { raw_output: rawOutput });
}

export async function saveGenerationRun(
  noteId: number,
  scene: string,
  stage: string,
  inputJson: string,
  promptFull: string,
): Promise<number> {
  return await invoke<number>('save_generation_run', {
    note_id: noteId,
    scene,
    stage,
    input_json: inputJson,
    prompt_full: promptFull,
  });
}

export async function listGenerationRuns(noteId: number): Promise<GenerationRun[]> {
  return await invoke<GenerationRun[]>('list_generation_runs', { note_id: noteId });
}

export async function updateGenerationOutput(
  runId: number,
  outputRaw: string,
  outputJson: string,
): Promise<void> {
  return await invoke<void>('update_generation_output', {
    run_id: runId,
    output_raw: outputRaw,
    output_json: outputJson,
  });
}

export async function markGenerationAdopted(runId: number): Promise<void> {
  return await invoke<void>('mark_generation_adopted', { run_id: runId });
}

// ── Style Profile ──

export async function extractStyleProfile(
  bookId: number,
  name: string,
  sourceScope: string,
  language: string,
): Promise<string> {
  return await invoke<string>('extract_style_profile', {
    book_id: bookId,
    name,
    source_scope: sourceScope,
    language,
  });
}

export async function getStyleProfile(bookId: number): Promise<StyleProfile | null> {
  return await invoke<StyleProfile | null>('get_style_profile', { book_id: bookId });
}

export async function listStyleProfiles(bookId: number): Promise<StyleProfile[]> {
  return await invoke<StyleProfile[]>('list_style_profiles', { book_id: bookId });
}

export async function getStyleExamples(profileId: number): Promise<StyleExample[]> {
  return await invoke<StyleExample[]>('get_style_examples', { profile_id: profileId });
}

export async function deleteStyleProfile(profileId: number): Promise<void> {
  return await invoke<void>('delete_style_profile', { profile_id: profileId });
}

File: src-tauri/src/commands/ai.rs
use crate::services::generation::{self, WritingTask, StructuredOutput};
use crate::state;

#[tauri::command]
pub(crate) async fn assemble_ai_prompt(note_id: i64, task: WritingTask) -> Result<String, String> {
    generation::assemble_prompt(note_id, &task).await
}

#[tauri::command]
pub(crate) async fn parse_ai_output(raw_output: String) -> Result<StructuredOutput, String> {
    Ok(generation::parse_structured_output(&raw_output))
}

#[tauri::command]
pub(crate) async fn save_generation_run(
    note_id: i64,
    scene: String,
    stage: String,
    input_json: String,
    prompt_full: String,
) -> Result<i64, String> {
    let db = state::get_db().await?;
    let now = chrono::Utc::now().to_rfc3339();
    let run = crate::db::GenerationRun {
        id: 0,
        note_id,
        scene,
        stage,
        input_json,
        prompt_full,
        output_raw: None,
        output_json: None,
        user_adopted: 0,
        created_at: now,
    };
    db.insert_generation_run(&run).await.map_err(|e| e.to_string())
}

#[tauri::command]
pub(crate) async fn list_generation_runs(note_id: i64) -> Result<Vec<crate::db::GenerationRun>, String> {
    let db = state::get_db().await?;
    db.list_generation_runs_by_note(note_id).await.map_err(|e| e.to_string())
}

#[tauri::command]
pub(crate) async fn update_generation_output(
    run_id: i64,
    output_raw: String,
    output_json: String,
) -> Result<(), String> {
    let db = state::get_db().await?;
    db.update_generation_output(run_id, &output_raw, &output_json).await.map_err(|e| e.to_string())
}

#[tauri::command]
pub(crate) async fn mark_generation_adopted(run_id: i64) -> Result<(), String> {
    let db = state::get_db().await?;
    db.mark_generation_adopted(run_id).await.map_err(|e| e.to_string())
}
